# Real Dataset Load

In [1]:
import pandas as pd
from sklearn.datasets import load_diabetes

# ১. Real diabetes data load korlam
diabetes = load_diabetes(as_frame=True)
df_diabetes = diabetes.frame

# ২. Dataset-er prothom 5 ti row dekhi
print(df_diabetes.head())


        age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  target  
0 -0.002592  0.019907 -0.017646   151.0  
1 -0.039493 -0.068332 -0.092204    75.0  
2 -0.002592  0.002861 -0.025930   141.0  
3  0.034309  0.022688 -0.009362   206.0  
4 -0.002592 -0.031988 -0.046641   135.0  


# Convert Text to Number

In [2]:
# Dummy context-er jonno ekti City column jog korlam (Text data)
df_diabetes['City'] = ['Dhaka', 'Chittagong', 'Sylhet', 'Dhaka', 'Sylhet'] * 88 + ['Dhaka', 'Chittagong']

# Text column-ke one-hot encoding korar standard Pandas function:
df_encoded = pd.get_dummies(df_diabetes, columns=['City'], drop_first=True)

print("\nEncoding korar por nuthun columns gulo dekho:")
print(df_encoded.head())



Encoding korar por nuthun columns gulo dekho:
        age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  target  City_Dhaka  City_Sylhet  
0 -0.002592  0.019907 -0.017646   151.0        True        False  
1 -0.039493 -0.068332 -0.092204    75.0       False        False  
2 -0.002592  0.002861 -0.025930   141.0       False         True  
3  0.034309  0.022688 -0.009362   206.0        True        False  
4 -0.002592 -0.031988 -0.046641   135.0       False         True  


# Target Split,Training & R^2 Score Test

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# ১. X and y separation
# Target column bad diye baki shob column holo X
X_real = df_encoded.drop(columns=['target'])
y_real = df_encoded['target']

# ২. Train-Test Split (80% Train, 20% Test)
X_train_r, X_test_r, y_train_r, y_test_r = train_test_split(X_real, y_real, test_size=0.2, random_state=42)

# ৩. Model setup & training
real_model = LinearRegression()
real_model.fit(X_train_r, y_train_r)

# ৪. Prediction & R2 Score Calculation
real_predictions = real_model.predict(X_test_r)
real_r2 = r2_score(y_test_r, real_predictions)

print(f"\nReal Diabetes Dataset-e Model-er R2 Score: {real_r2:.4f}")



Real Diabetes Dataset-e Model-er R2 Score: 0.4546


# Tuning solution

In [4]:
from sklearn.tree import DecisionTreeRegressor # <- Regressor tree imported
from sklearn.metrics import r2_score

# ১. Model Object setup
tree_reg = DecisionTreeRegressor(random_state=42)

# ২. Model Training phase
tree_reg.fit(X_train_r, y_train_r)

# ৩. Testing Exam & Scoring
tree_pred = tree_reg.predict(X_test_r)
tree_r2 = r2_score(y_test_r, tree_pred)

print(f"Decision Tree Regressor er R2 Score: {tree_r2:.4f}")


Decision Tree Regressor er R2 Score: 0.0100


In [ ]:
# tree-r depth limit kore dilam max_depth=3 diye
tuned_tree = DecisionTreeRegressor(max_depth=3, random_state=42)

# Re-train 
tuned_tree.fit(X_train_r, y_train_r)

# Re-test and Evaluate
tuned_pred = tuned_tree.predict(X_test_r)
tuned_r2 = r2_score(y_test_r, tuned_pred)

print(f"Tuned Decision Tree er R2 Score: {tuned_r2:.4f}")


Tuned Decision Tree er R2 Score: 0.3294


# Project Conclusion & Key Takeaways:



1.⁠ ⁠Objective Achievement
In this project, an end-to-end Machine Learning Regression pipeline was successfully implemented to predict diabetes progression scores using real-world patient features (such as Age, BMI, and Blood Pressure). As part of the data preprocessing phase, categorical string inputs were successfully transformed into numerical formats using One-Hot Encoding (⁠pd.get_dummies⁠), satisfying the structural requirements of Scikit-Learn models.

2.⁠ ⁠Model Performance & Overfitting Analysis
The evaluation phase involved comparing two distinct predictive algorithms based on their R^2 Score (Coefficient of Determination) metrics:

 Linear Regression (Baseline): Despite encountering real-world noisy parameters, this model successfully captured 45.46% (R^2 = 0.4546) of the data variance, serving as a reliable linear baseline.
 Decision Tree Regressor (Default): The unconstrained model fell into a severe Overfitting trap, with its test accuracy collapsing to 1.00% (R^2 = 0.0100). This occurred because the algorithm completely memorized the training dataset rows instead of extracting generalized patterns, failing significantly when exposed to unseen test data.


 3.⁠ ⁠Hyperparameter Tuning Solution
To resolve the overfitting challenge, manual Hyperparameter Tuning was applied by restricting the structural expansion of the flowchart paths. Limiting the tree depth via the ⁠max_depth=3⁠ parameter forced the model to ignore micro-level noise and prioritize the top 3 structural data patterns. This constraint instantly optimized the model's generalization capabilities, lifting the R^2 score from a critical 1.00% to 32.94% (R^2 = 0.3294).

4.⁠ ⁠Future Recommendation
To elevate the predictive accuracy further, it is recommended to transition toward ensemble machine learning frameworks—specifically a Random Forest Regressor. Combining predictions across multiple randomized decision trees will naturally mitigate high variance and stabilize performance outputs on complex, noisy distributions...
